Document difference of C++ for OpenCL wrt OpenCL C (#236)
KhronosGroup/OpenCL-Docs: cxx4opencl/diff2openclc.txt

> === Difference to OpenCL C

Reviewer: I've no idea if this is a requirement but other files seem to have a copyright header.

> {cpp} for OpenCL targets backwards compatibility to OpenCL C

Reviewer 2: ```suggestion
{cpp} for OpenCL aims to achieve backwards compatibility with OpenCL C
```

> in the majority of features. However, it is a different

Reviewer 2: ```suggestion
for the majority of features. However, it is a different
```

language that is derived from {cpp} and therefore it inherits

> its fundamental design principles. Hence {cpp} for OpenCL will

Reviewer 2: ```suggestion
its fundamental design principles. Hence {cpp} for OpenCL
```

> deviate from OpenCL C in the same areas where {cpp} deviates

Reviewer 2: ```suggestion
deviates from OpenCL C in the same areas where {cpp} deviates
```

from C.

In this section some of the most common areas where it is

> expected to behave differently are described.

Reviewer 2: ```suggestion
This section describes the main differences between {cpp} for
OpenCL and OpenCL C.
```

==== Implicit Conversions

> {cpp} is much more strict about conversions between types

Reviewer 2: ```suggestion
{cpp} is much stricter about conversions between types,
```

especially those that are performed implicitly by the compiler.

> For example it is not possible to convert non-`const` object

Reviewer 2: ```suggestion
For example it is not possible to convert a non-`const` object
```

> to `const` explicitly. For details please refer to {cpp}17

Reviewer: I believe the terms in this sentence are swapped.

```suggestion
For example it is not possible to convert a `const` object
to non-`const` implicitly. For details please refer to {cpp}17
```

I'm also suggesting using _implicitly_ as it might be possible to use `const_cast` for this in some specific cases (which is outside the scope of this doc, I think).

(`ISO/IEC 14882:2017`) `[conv]`.

[source,c]
----------
void foo(){
  const int* ptrconst;

>   int* ptr = ptrconst; // rejected in C++ for OpenCL

Reviewer 2: ```suggestion
  const int *ptrconst;
  int *ptr = ptrconst; // rejected in C++ for OpenCL
```

}
----------

> The same situation applies to narrowing conversions in

Reviewer 2: ```suggestion
The same applies to narrowing conversions in
```

initialization lists ({cpp}17 `[dcl.init.list]`). 

[source,c]
----------
struct mytype {
 int i;
};
void foo(uint par){
  mytype var = {

>     .i = par

Reviewer 2: It would help if you add a comment to highlight the issue.

Reviewer: I would suggest adding an example such as `(int2)(4.2f, 3.14f)` as well.

Author: Btw I realized that this rule is just for the initialization list but I am not sure vector initialization falls under this category. For now I will avoid adding a vector example until I investigate this further.

  };
}
----------

Note that some compilers will allow to silence this error
using some flags (e. g. in Clang `-Wno-error=c++11-narrowing`

> can be used).

Reviewer 2: ```suggestion
Some compilers allow silencing this error using a flag
(e.g. in Clang `-Wno-error=c++11-narrowing` can be used).
```

> Among other common conversions that won't be compiled in C++

Reviewer 2: ```suggestion
Among other common conversions that will not be compiled in C++
```

mode there are pointer to integers or integer to pointer

> conversions.

Reviewer: ```suggestion
mode there are pointer to integer or integer to pointer
type conversions.
```

[source,c]
----------
void foo(){

>   int* ptr;

Reviewer 2: ```suggestion
  int *ptr;
```

>   int i = ptr;

Reviewer 2: Again, would be useful to highlight the issue with a comment.

}
----------

==== Null Literal

> In C and OpenCL C null literal is being defined using other

Reviewer 2: ```suggestion
In C and OpenCL C the null literal is defined using other
```

language features as it is not represented explicitly i.e.
commonly it is defined as

[source,c]
----------
#define NULL (void*)0
----------

In {cpp} there is an explicit builtin literal `nullptr`
that should be used instead ({cpp}17 `[lex.nullptr]`).

{cpp} for OpenCL does not define `NULL` and therefore any
source code using it must be modified to use `nullptr`
instead. However as a workaround to avoid large modifications
`NULL` can also be defined/aliased to `nullptr` in custom
headers or using command line flag `-D`. It is not recommended

> to reuse C definition of `NULL` in {cpp} for OpenCL as it may

Reviewer 2: ```suggestion
to reuse the C definition of `NULL` in {cpp} for OpenCL as it may
```

cause compilation failures in cases that work for C.

[source,c]
----------
void foo(){

>  int* ptr = NULL; // inivalid initialization of int* with void*

Reviewer 2: ```suggestion
 int *ptr = NULL; // invalid initialization of int* with void*
```

}
----------

==== Use of restrict

> C++17 does not support `restrict` and therefore {cpp} for OpenCL

Reviewer: I'm not familiar with the markup language used in this file, so maybe it's all fine. I've noticed you use `{cpp}` everywhere but here you use `C++17`. Should this be replaced by `{cpp}17` or something?

```suggestion
{cpp}17 does not support `restrict` and therefore {cpp} for OpenCL
```

can not support it either. Some compilers might provide extensions

> with some functionality of `restrict` in {cpp} e. g. `__restrict`

Reviewer 2: ```suggestion
with some functionality of `restrict` in {cpp}, e.g. `__restrict`
```

in Clang.

==== Limitations of goto statements

{cpp} is more restrictive with respect to entering the scope of
variables than C. It is not possible to jump forward over a variable
declaration statement apart from some exceptions detailed in {cpp}17
`[stmt.dcl]`.

[source,c]
----------
if (cond)
  goto label;
int n = foo();
label:  // invalid: jumping forward over declaration of n
   // ...
----------

==== Use of Clang Blocks

> Clang Blocks that are defined by ObjC language are not supported

Reviewer: ```suggestion
Clang Blocks that are defined by the Objective-C language are not supported
```

and their use can be replaced by lambdas ({cpp}17 `[expr.prim.lambda]`).

This means that builtin function `enqueue_kernel` with OpenCL C

> prototype `OpenCL v2.0 s6.13.17` is not supported in {cpp} for OpenCL.

Reviewer: There are other BIF that rely on blocks, possibly in extensions. I would suggest generalising this statement a bit, maybe as follows: 

```suggestion
This means that builtin functions using blocks, such as `enqueue_kernel`,
are not supported in {cpp} for OpenCL.
```

Author: Just FYI I would still like to enumerate those from the core spec here.

// http://creativecommons.org/licenses/by/4.0/

=== Difference to OpenCL C

{cpp} for OpenCL aims to achieve backwards compatibility with OpenCL C
for the majority of features. However, it is a different
language that is derived from {cpp} and therefore it inherits
its fundamental design principles. Hence {cpp} for OpenCL
deviates from OpenCL C in the same areas where {cpp} deviates
from C.

This section describes the main differences between {cpp} for
OpenCL and OpenCL C.

==== Implicit Conversions

{cpp} is much stricter about conversions between types,
especially those that are performed implicitly by the compiler.
For example it is not possible to convert a `const` object
to non-`const` implicitly. For details please refer to {cpp}17
(`ISO/IEC 14882:2017`) `[conv]`.

[source,c]
----------
void foo(){
  const int *ptrconst;
  int *ptr = ptrconst; // invalid initialization discards const qualifier
}
----------

The same applies to narrowing conversions in
initialization lists ({cpp}17 `[dcl.init.list]`). 

[source,c]
----------
struct mytype {
 int i;
};
void foo(uint par){
  mytype var = {
    .i = par // narrowing from uint to int is disallowed
  };
}

> ----------

Reviewer 3: This example should be simplified. The structure isn't really needed.

Author: Hmm... I don't know how to show it without... Can you paste an example you have in mind?

Reviewer 3: Oh, I'd read a bit too quickly. The restriction only applies to initialisation lists. The example is fine as is.

----------
void foo(){
  int *ptr;
  int i = ptr; // incompatible pointer to integer conversion
}
----------

==== Null Literal

In C and OpenCL C the null literal is defined using other
language features as it is not represented explicitly i.e.
commonly it is defined as

[source,c]
----------
#define NULL (void*)0
----------

In {cpp} there is an explicit builtin literal `nullptr`
that should be used instead ({cpp}17 `[lex.nullptr]`).

{cpp} for OpenCL does not define `NULL` and therefore any
source code using it must be modified to use `nullptr`
instead. However as a workaround to avoid large modifications
`NULL` can also be defined/aliased to `nullptr` in custom
headers or using command line flag `-D`. It is not recommended
to reuse the C definition of `NULL` in {cpp} for OpenCL as it may
cause compilation failures in cases that work for C.

[source,c]
----------
void foo(){
 int *ptr = NULL; // invalid initialization of int* with void*
}
----------

==== Use of restrict

{cpp}17 does not support `restrict` and therefore {cpp} for OpenCL
can not support it either. Some compilers might provide extensions
with some functionality of `restrict` in {cpp}, e.g. `__restrict`

> in Clang.

Reviewer 3: Maybe worth a comment explaining that defining `restrict` to an empty macro or `__restrict` is possible.

Author: I don't like the official language documentation is the best place to describe workarounds like this... Besides, it is pretty straight forward to get.

Author: Also developers might prefer to alias it to __restrict or modify source code. There can be many approaches depending on the situation.

Reviewer 3: Well, we've done it for `NULL`...

>developers might prefer to alias it to __restrict or modify source code

Yes, as said in my comment.

Author: Fair point! I have added a paragraph explaining that... although generally I think we should not go too much in this direction...

In C and OpenCL C the null literal is defined using other
language features as it is not represented explicitly i.e.
commonly it is defined as

[source,c]
----------
#define NULL (void*)0
----------

In {cpp} there is an explicit builtin literal `nullptr`
that should be used instead ({cpp}17 `[lex.nullptr]`).

{cpp} for OpenCL does not define `NULL` and therefore any
source code using it must be modified to use `nullptr`
instead. However as a workaround to avoid large modifications
`NULL` can also be defined/aliased to `nullptr` in custom
headers or using command line flag `-D`. It is not recommended
to reuse the C definition of `NULL` in {cpp} for OpenCL as it may
cause compilation failures in cases that work for C.

[source,c]
----------
void foo(){
 int *ptr = NULL; // invalid initialization of int* with void*
}
----------

==== Use of restrict

{cpp}17 does not support `restrict` and therefore {cpp} for OpenCL
can not support it either. Some compilers might provide extensions
with some functionality of `restrict` in {cpp}, e.g. `__restrict`
in Clang.

This feature only affects optimizations and the source
code can be modified by removing it. As a workaround to avoid manual

> modifications macro substitutions can be used to either remove the

Reviewer 2: ```suggestion
modifications, macro substitutions can be used to either remove the
```
